### ISSUE TITLE
Feature Gates 

### ISSUE THREAD

**Issue by erictune** [kubernetes-sigs/kubebuilder#849]
Core Kubernetes types have Feature Gates. 
They are documented here: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

This is a good practice for anyone, whether they are building a CRD as part of the K8s org, or their own API.

Sadly, CRDs don't have feature gate support built into them.

KubeBuilder could close this gap.  It could provide a feature-gate feature.  
It might work like this:
* Developer tags a field in an API as belonging to a feature gate, like `// +feature-gate the-new-hotness
* KB (or controller-runtime or controller-tools) makes your controller binary have a `--feature-gates` flag.
* all "gate names" mentioned in tags are added as valid comma-separated arguments to `--feature-gates` flag. 
* When installing the CRD, controller-tools will blank out disabled parts of the schema.  This causes pruning to prune them.
* Generated documentation lists the supported feature gate strings.

/kind feature

**Comment by camilamacedo86** [comment]
It is related to: https://github.com/kubernetes-sigs/controller-tools/issues/600
Also, if you would like to understand more about the use cases see: https://www.youtube.com/watch?v=AJa2B-twtG4

**Comment by camilamacedo86** [comment]
This is an amazing RFE.

That said, I think this kind of feature would better belong in `controller-gen`.

We could potentially support markers to generate CRDs, RBAC, and other manifests conditionally based on feature gates (like `alpha`, `beta`, etc). That would allow more flexibility while keeping the code clean and consistent.

Since this relates directly to how manifests are generated, I don’t see how it would fit into Kubebuilder itself before `controller-gen` first implemented. 

Looping in @JoelSpeed and @sbueringer for thoughts:  
WDYT? Would it make sense to move this request to the `controller-gen` repo?

**Comment by JoelSpeed** [comment]
We have a pattern in openshift APIs where we have a custom marker built into controller-gen that we can use to annotate different versions of CRDs based on which gates are and aren't enabled. The process is quite involved and I'd definitely be keen to see something simpler evolve as a first class citizen of controller-gen.

However, there is a lot of nuance to feature gates.

For example, we talk in the initial description about different sections of the schema being blanked out and therefore those parts being pruned. I believe if you were using a K8s core API and tried to set a feature gate field, it would actually reject the request, so to be consistent, we would want to achieve something similar.

Secondly, gates create combinational issues. We need the ability to arbitrarily select a set of gates and generate the correct manifest, and then ensure that this version of the CRD is deployed correctly with the controller also specifying the same gates.

As the number of gates grows, the possible combinations of features will grow exponentially.

Then you have to factor that different existing markers also contribute to feature gates. For example, we might want to have feature gated XValIdation, since it only makes sense once a particular gate is enabled (or actually, pretty much any marker).
Once you get to this point, take for example of a feature gated enum marker, what happens if multiple gates affect the same enum? How do we merge the requirements? Or do we need to specify the enum marker that should take precedence when multiple gates are enabled?

There's a lot to think about here, but I think first up I'd be trying to work out what the end goal is for configuring which gates are/aren't enabled and how we make sure that the CRD and controller enablement are consistent.

**Comment by camilamacedo86** [comment]
Yes, I agree — it gets complex when multiple feature gates are involved.

For OLM v1, we decided to simplify things by offering just stable and experimental versions. We're using `controller-gen` to generate markers, and by limiting the gates to categories like `alpha` and `stable`, it becomes easier for end users. It also simplifies our CI — we can run e2e tests for all gates, or just for stable ones.

As an example, the OLM v1 CRD generator code is here:  
https://github.com/operator-framework/operator-controller/tree/main/hack/tools/crd-generator

We found the OCP downstream implementation a bit too complex to maintain, so this simpler approach might work well for `controller-gen`/Kubebuilder too.

I think the first step is enabling support in `controller-gen`. Once that’s in place, we could expose something like `create api --feature-gate` and build logic on top of it to streamline things further.

We’d also need to adapt the scaffolding and think about how to best support and simplify this use case.

**Comment by sbueringer** [comment]
Not sure if it's a good idea to end up with CRDs that don't match the Go types used in controllers & webhooks.

Or are you also modifiying controller-code based on the gates?

**Comment by camilamacedo86** [comment]
>Not sure if it's a good idea to end up with CRDs that don't match the Go types used in controllers & webhooks.
> Or are you also modifiying controller-code based on the gates?

The idea is to have markers in the go types that allow us to work with feature gates.
example

```
	// +kubebuilder:validation:MaxLength:=63
	// +kubebuilder:feature-gate- alpha 
	// +required
	Namespace string `json:"namespace"`
```


That means controller-gen will not render the spec  Namespace in the stable CRD
But will handle the spec  Namespace in the Alpha CRD.

**Comment by JoelSpeed** [comment]
Typically you would also have something in the controller that would gate functionality based on a particular feature gate as well.

I think it's important to note here that gates are for specific features, the `alpha`, `beta`, `GA` state is independent of the actual gates, I would call these more feature sets, so perhaps somehow we need to define also feature sets that contain gates?

**Comment by camilamacedo86** [comment]
Yes, the idea is to allow **Go type markers** to define which fields are tied to specific feature gates.

For example:

```go
// +kubebuilder:validation:MaxLength=63
// +kubebuilder:feature-gate=alpha
// +kubebuilder:validation:Required
Namespace string `json:"namespace,omitempty"`
``` 

With this, `controller-gen` could generate:

- A **"stable" CRD** where `spec.namespace` is completely omitted.
- An **"alpha" CRD** where `spec.namespace` is included and validated.

This is useful when a new feature (e.g., Feature Gate `A`) introduces new API fields. We don't want those fields exposed in the default/stable CRD until the feature is ready — but we **do** want the ability to test and enable them behind a gate.

The same principle would apply to **RBAC** and other generated manifests too.

So yes, the goal is to **control CRD output via feature gates**, without necessarily needing to fork or rewrite the Go type definitions — keeping the controller and the API aligned, but with flexibility in what gets exposed to users based on the feature state.

I created an RFE for controller-gen: https://github.com/kubernetes-sigs/controller-tools/issues/1238 ( so that we can discuss the specific part for controller-tools ). If I accept, then Kubebuilder can see how to handle the scenarios and help users. 


This one is blocked by: https://github.com/kubernetes-sigs/controller-tools/issues/1238
